PyRoute/Calculation/TradeCalculation.py

```diff
@@ -68,23 +69,33 @@ def __init__(self, galaxy, min_btn=13, route_btn=8, route_reuse=10, debug_flag=F
         # Are debugging gubbins turned on?
         self.debug_flag = debug_flag
 
-        self.shortest_path_tree = None
-
     def base_route_filter(self, star, neighbor):
-        if star in self.redzone or neighbor in self.redzone:
```

Comment:
Replace the redzone/greenzone sets in the TradeCaclulation class with a "redzone" flag on each star. Set the flag correctly in the setup for the Star (when read and everything else it tested). This whole method becomes `return star.redzone or neighbor.redzone`

Reply:
Will do that, didn't want to load too many new flags up.  How does your suggestion work for XRouteCalculation, CommCalculation, etc?

Reply:
Works the same way in `XRouteCalculation` and `CommCalculations` but they also want to know who is in the same allegiance code so the communications routes don't cross borders. The `AllyGen.are_allies(...)` can be otherwise left. The number of routes these two route calculators process is orders of magnitude smaller, so don't worry about the hard core optimizations.

Reply:
Hang on - that early in the piece, the galaxy.trade object is not yet set, so it doesn't look like we know what criterion to apply to set redzone.

`{Comm/XRoute/Trade}Calculation` all exclude actually-red-zoned systems (zone codes R or F), so could set on that alone, and then take advantage of that in the itertools.combinations call, via yet another bloody comprehension:
```
    def generate_base_routes(self):
        self.logger.info('generating jumps...')
        self.galaxy.is_well_formed()
        for star, neighbor in itertools.combinations(self.galaxy.ranges, 2):
```

becomes something like:
```
    def generate_base_routes(self):
        self.logger.info('generating jumps...')
        self.galaxy.is_well_formed()
        raw_ranges = ( star, neighbour for star, neighbour in itertools.combinations(self.galaxy.ranges, 2) if not star.redzone and not neighbour.redzone )
        for star, neighbor in raw_ranges:
```

Reply:
The unilateral-filtering gubbins hooks up the trade object early enough to make this work, and sets redzone according to the particular object's rules (eg TradeCalculation, CommCalculation, etc).